modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/GlobalSnapshotStateChannelEventsProcessor.scala

```diff
@@ -67,26 +66,63 @@ object GlobalSnapshotStateChannelEventsProcessor {
     new GlobalSnapshotStateChannelEventsProcessor[F] {
       private val logger = Slf4jLogger.getLoggerFromClass[F](GlobalSnapshotStateChannelEventsProcessor.getClass)
 
+      def deserialize[A: Decoder](binary: Signed[StateChannelSnapshotBinary]): F[Option[A]] =
+        jsonBrotliBinarySerializer.deserialize[A](binary.value.content).map(_.toOption)
+
+      def buildSnapshotFeesInfo(
+        lastGlobalSnapshotInfo: GlobalSnapshotInfo,
+        event: StateChannelOutput,
+        allFeesAddresses: Map[Address, Set[Address]]
+      ): F[SnapshotFeesInfo] =
+        event.snapshotBinary.value.lastSnapshotHash match {
+          case hash if hash == Hash.empty => SnapshotFeesInfo.empty.pure // genesis
+          case _ =>
+            deserialize[Signed[CurrencyIncrementalSnapshot]](event.snapshotBinary).flatMap {
+              case None => new Exception("Could not get snapshot after deserializing").raiseError[F, SnapshotFeesInfo]
+              case Some(snapshot) =>
+                Async[F].delay {
+                  val stakingBalance = fetchStakingBalance(event.address, lastGlobalSnapshotInfo)
+                  val maybeOwnerAddress = snapshot.value.messages.flatMap(_.find(_.messageType == MessageType.Owner)).map(_.address)
+                  val maybeStakingAddress = snapshot.value.messages.flatMap(_.find(_.messageType == MessageType.Staking)).map(_.address)
```

Comment:
`snapshot.value.messages` are `SortedSet` which indicates that we can have `n` messages there (not only 2 messages). If only 1 staking and 1 owner are allowed then I am wondering why `messages` in `CurrencyIncrementalSnapshot` are not something like: `case class Messages(Option[Owner], Option[Staking])`.

But if `SortedSet` was used on purpose (so we allow `n` messages) then I would be careful with `find` as if finds the very first message of a specific type.

If `SortedSet` uses `Ordering` based on ordinal ASC then it will return the one with lowest ordinal (reverse needed to `findLast`)
If `SortedSet` uses `Ordering` based on ordinal DESC then it will return the one with highest ordinal
If `SortedSet` uses `Ordering` which is not specified then it is kinda unpredictable and I would `maxBy(_.ordinal)` or sth.

Reply:
Of course, the ultimate solution would be to provide proper `Ordering` to `SortedSet` globally based on `ordinal` because sorting `SortedSet` manually makes `SortedSet` useless.